esp8266/esp8266.go

```diff
@@ -0,0 +1,249 @@
+// Package esp8266 implements TCP/UDP communication over serial
+// with a separate Wifi ESP8266 board using the Espressif AT command set
+// across a UART interface.
+//
+// In order to use this driver, the ESP8266 must be flashed with firmware
+// supporting the AT command set. Many ESP8266 chips already have this firmware
+// installed by default. You will need to install this firmware if you have an
+// ESP8266 that has been flashed with NodeMCU (Lua) or Arduino firmware.
+//
+// Datasheet:
+// https://www.espressif.com/sites/default/files/documentation/0a-esp8266ex_datasheet_en.pdf
+//
+// AT command set:
+// https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf
+//
+package esp8266
+
+import (
+	"machine"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Device wraps UART connection to the ESP8266.
+type Device struct {
+	bus machine.UART
+
+	// command responses that come back from the ESP8266
+	response []byte
+
+	// data received from a TCP/UDP connection forwarded by the ESP8266
+	socketdata []byte
+}
+
+// New returns a new esp8266-wifi driver. Pass in a fully configured UART bus.
+func New(b machine.UART) *Device {
+	return &Device{bus: b, response: make([]byte, 512), socketdata: make([]byte, 0, 1024)}
+}
+
+// Configure sets up the device for communication.
+func (d Device) Configure() {
+}
+
+// Connected checks if there is communication with the ESP8266.
+func (d *Device) Connected() bool {
+	d.Execute(Test)
+
+	// handle response here, should include "OK"
+	r := d.Response()
+	if strings.Contains(string(r), "OK") {
+		return true
+	}
+	return false
+}
+
+// Write raw bytes to the UART.
+func (d *Device) Write(b []byte) (n int, err error) {
+	return d.bus.Write(b)
+}
+
+// Read raw bytes from the UART.
+func (d *Device) Read(b []byte) (n int, err error) {
+	return d.bus.Read(b)
+}
+
+// how long in milliseconds to pause after sending AT commands
+const pause = 100
+
+// Execute sends an AT command to the ESP8266.
+func (d Device) Execute(cmd string) error {
+	_, err := d.Write([]byte("AT" + cmd + "\r\n"))
+	return err
+}
+
+// Query sends an AT command to the ESP8266 that returns the
+// current value for some configuration parameter.
+func (d Device) Query(cmd string) (string, error) {
+	_, err := d.Write([]byte("AT" + cmd + "?\r\n"))
+	return "", err
+}
+
+// Set sends an AT command with params to the ESP8266 for a
+// configuration value to be set.
+func (d Device) Set(cmd, params string) error {
+	_, err := d.Write([]byte("AT" + cmd + "=" + params + "\r\n"))
+	return err
+}
+
+// Version returns the ESP8266 firmware version info.
+func (d Device) Version() []byte {
+	d.Execute(Version)
+	return d.Response()
+}
+
+// Echo sets the ESP8266 echo setting.
+func (d Device) Echo(set bool) {
+	if set {
+		d.Execute(EchoConfigOn)
+	} else {
+		d.Execute(EchoConfigOff)
+	}
+	// TODO: check for success
+	d.Response()
```

Comment:
There are many places like this, that seem to ignore any potential errors.
Is this intentional or just not yet handled? I can imagine that it may be difficult to handle errors and even harder to recover from them.

Reply:
The error handling is still to be added in a future iteration. We have some improvements to the parsing and string handing now in TinyGo itself that should make this a bit easier. 

However, this method should probably return a `nil` even if the error handling is incomplete just to  avoid changing the API.